### PR TITLE
Python: Add support for magic methods

### PR DESCRIPTION
**Pull request by yoff**
This PR labels certain control flow nodes as corresponding to method calls to magic methods. This is then picked up by the data flow library to resolve flow between subexpressions and parameters.

Currently only binary operators are supported and only the standard form (not the r-variant).

### CODE REVIEW

**Comment by yoff** [comment]
Ah, I forgot to annotate the test file with the newly obtained results..

**Comment by yoff** [comment]
There were no annotations present, so I did it in two steps to make the diff clear. The first commit marks all calls as not found, the second one unmarks the ones found by the new magic feature.

File: python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll
Comment (intrigus-lgtm) [inline python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll:172]:
```suggestion
  /** Get the specified argument to this call. */
```

File: python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll
Comment (intrigus-lgtm) [inline python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll:183]:
```suggestion
/** Represents a call to a callable. */
```

File: python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll
Comment (intrigus-lgtm) [inline python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll:206]: Missing QlDoc?

File: python/ql/src/semmle/python/Magic.qll
Comment (intrigus-lgtm) [inline python/ql/src/semmle/python/Magic.qll:19]:
```suggestion
    /** Gets the name of the method that would be called. */
```

File: python/ql/src/semmle/python/Magic.qll
Comment (yoff) [reply python/ql/src/semmle/python/Magic.qll]: It has been remedied :-)

File: python/ql/src/semmle/python/Magic.qll
Comment (yoff) [reply python/ql/src/semmle/python/Magic.qll]: I have adopted the above structure.

File: python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll
Comment (yoff) [reply python/ql/src/experimental/dataflow/internal/DataFlowPrivate.qll]: Yes, added.

File: python/ql/src/semmle/python/SpecialMethods.qll
Comment (tausbn) [inline python/ql/src/semmle/python/SpecialMethods.qll:13]:
```suggestion
/** A control flow node which might correspond to a special method call. */
```

File: python/ql/src/semmle/python/SpecialMethods.qll
Comment (tausbn) [inline python/ql/src/semmle/python/SpecialMethods.qll:23]:
```suggestion
  /** A control flow node which might correspond to a special method call. */
```

File: python/ql/src/semmle/python/SpecialMethods.qll
Comment (tausbn) [inline python/ql/src/semmle/python/SpecialMethods.qll:28]:
```suggestion
    /** Gets the control flow node that would be passed as the specified argument. */
```

**Review by tausbn — CHANGES_REQUESTED** [review]
Found a few typos. Otherwise looks good to me. :+1:

**Review by tausbn — APPROVED** [review]

**Comment by tausbn** [comment]
I think this looks good to go, once the merge conflicts have been ironed out.

**Comment by yoff** [comment]
"dismissed"?! I Specifically merged main to iron out those conflicts...

**Comment by RasmusWL** [comment]
Errors on jenkins `ERROR: The cast type DataFlowCall is incompatible with the expression type ControlFlowNode. (argumentRouting1.ql:12,23-45)`

**Review by tausbn — APPROVED** [review]

**Pull request merged**
